internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go

```diff
@@ -46,19 +55,116 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
 		return nil, errors.New(err, "failed to unpack configuration")
 	}
 	return &contextProviderK8sSecrets{
-		logger: logger,
-		config: &cfg,
+		logger:       logger,
+		config:       &cfg,
+		secretsCache: make(map[string]secretsData),
 	}, nil
 }
 
 func (p *contextProviderK8sSecrets) Fetch(key string) (string, bool) {
-	// key = "kubernetes_secrets.somenamespace.somesecret.value"
+	return p.getFromCache(key)
+}
+
+// Run initializes the k8s secrets context provider.
+func (p *contextProviderK8sSecrets) Run(ctx context.Context, comm corecomp.ContextProviderComm) error {
+	client, err := getK8sClientFunc(p.config.KubeConfig, p.config.KubeClientOptions)
+	if err != nil {
+		p.logger.Debugf("Kubernetes_secrets provider skipped, unable to connect: %s", err)
+		return nil
+	}
+	p.clientMx.Lock()
+	p.client = client
+	p.clientMx.Unlock()
+	go p.updateSecrets(ctx)
+
+	<-comm.Done()
+
+	p.clientMx.Lock()
+	p.client = nil
+	p.clientMx.Unlock()
+	return comm.Err()
+}
+
+func getK8sClient(kubeconfig string, opt kubernetes.KubeClientOptions) (k8sclient.Interface, error) {
+	return kubernetes.GetKubernetesClient(kubeconfig, opt)
+}
+
+// Update the secrets in the cache every TTL minutes
+func (p *contextProviderK8sSecrets) updateSecrets(ctx context.Context) {
+	timer := time.NewTimer(p.config.TTLUpdate)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			p.updateCache()
+			timer.Reset(p.config.TTLUpdate)
+		}
+	}
+}
+
+func (p *contextProviderK8sSecrets) updateCache() {
+	p.secretsCacheMx.Lock()
+	for name, data := range p.secretsCache {
+		diff := time.Since(data.lastAccess)
+		if diff > p.config.TTLDelete {
+			delete(p.secretsCache, name)
+		} else {
+			newValue, ok := p.fetchSecret(name)
```

Comment:
This is another issue that I am seeing. The lock is being held when making API calls to fetch the secret. What is the worry that this blocks for a long time holding the lock and then causing the rendering of the policy to be blocked?

Is it possible to be multiple key values in a single API call? It might be better to change this to first get all the keys that new values are needed. Then make a single API call to get all the values (if possible), and then update the cache with the values.

Reply:
> Is it possible to be multiple key values in a single API call? 

It is not possible. The only way to get multiple secrets at once is to use the list function instead of get [here](https://github.com/elastic/elastic-agent/blob/b3ba2192a841d4e353ac625620a88643b87144ef/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go#L202). However, the options to filter the secrets we return are [these](https://github.com/kubernetes/kubernetes/blob/ad9b60e2c9ddb21e8b00cabbe27e639638a0ea88/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L322), and as we can see, there are none to filter by the name. So in the end we would be retrieving all the secrets from a namespace when there is no need.

Reply:
Dang. Okay, I am still worried about it holding the lock the entire time it is refreshing all the secrets.

Reply:
Does `fetchSecrets` have a timeout? What if you never get a response? What if you do get a response, but it takes one hour?

Now that `cacheTmp` is being rebuilt is holding the lock for each `fetchSecret` call still necessary? Can `cacheTmp` be built and then assigned to `p.secretsCache` with only that one line holding the lock?

Since we are iterating over secretsCache that would likely require duplicating the keys first.

> However, the options to filter the secrets we return are these, and as we can see, there are none to filter by the name. So in the end we would be retrieving all the secrets from a namespace when there is no need.

What places more load on the k8s API? Returning a single response that is unnecessarily large or multiple small requests made rapidly one after another?

We should bias towards making sure the k8s api-server remains stable over optimizing the request size if possible.

Reply:
> Does fetchSecrets have a timeout? What if you never get a response? What if you do get a response, but it takes one hour?

I added a timeout just now, so we will never be waiting for 1h.

> Now that cacheTmp is being rebuilt is holding the lock for each fetchSecret call still necessary?

I changed the implementation to this way:
1. Created the temporary cache.
2. Apply read lock and iterate over the entries of the secrets cache. If entry remains, then write on temporary cache.
3. Now that the iteration is over, apply writing lock to secrets cache and make it the same as the temporary one.

Is this ok this way @cmacknz ?